Gutenberg - Audio Block Processor

jd-alexander opened the pull request:
## Findings
This PR is the Android side of the [iOS implementation for updating the HTML content](https://github.com/wordpress-mobile/WordPress-iOS/pull/15420) of the Audio Block without Gutenberg being open. 

## Solution
The `AudioBlockProcessor` replaces the `<audio src>` tag that contains local file URL with a remote media URL once the file upload is complete. The id of the Audio Block is also updated with a remote variant as well. 

## Testing

1. Add an Audio block to a new post.
2. Upload an audio file to the block (I  lessened the Cellular Network type and Signal strength to replicate a slow network so I  had enough time to close the editor).
3. Close the editor by pressing the back button.
4. Let the upload to finish in the background.
5. When the audio file is uploaded and the post saved, open the post again (on web or mobile).

Verify that the behavior is functioning correctly by :
  - Checking that the block renders properly.
  - Checking that all references to src are updated to the remote URL.
  - Checking that the media ID has been updated to the remote one.

## Reviewing 
* Only 1 reviewer is needed but anyone can review. 
* This PR depends on the [Consolidated Media Picker Audio Block Integration PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13612) so that has to be merged before this is merged in.

## Submitter Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] I have considered adding accessibility improvements for my changes.
- [x] If it's feasible, I have added unit tests.

jd-alexander committed:
Created the AudioBlockProcessor

jd-alexander committed:
Created AUDIO MediaBlockType.

jd-alexander committed:
BlockProcessorFactory is made aware of the AudioBlockProcessor

jd-alexander committed:
Created tests for the AudioBlockProcessor

loremattei commented:
Hey! I've moved this to 16.5 because 16.4 has been cut. If you want this to make it to 16.4, please feel free to ping me.

jd-alexander committed:
Merge branch 'gutenberg/audio_block_consolidated_media_picker_integration' into gutenberg/audio-block-processor

jd-alexander committed:
Updated ref.

jd-alexander committed:
Added newline to the AudioBlockProcessor

jd-alexander committed:
Added audio block to the patterns that check if gb media is in post body

jd-alexander committed:
Added unit tests for the Audio block's isMediaInGutenberg

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessor.kt:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    override fun processBlockJsonAttributes(jsonAttributes: JsonObject?): Boolean {
+        val id = jsonAttributes?.get(FileBlockProcessor.ID_ATTRIBUTE)
```
NP: We can use the local `ID_ATTRIBUTE` instead of the one from `FileBlockProcessor`

```suggestion
        val id = jsonAttributes?.get(ID_ATTRIBUTE)
```

jd-alexander replied:
whoops! Thanks for spotting that @mkevins 🙇

mkevins reviewed (approved):
Nice work @jd-alexander :+1: 

Code looks good, and thanks for writing the tests too! I've tested this on Pixel 3a (physical device) and it's working as expected.

I left one minor suggestion. Other than that, LGTM!

jd-alexander committed:
Update WordPress/src/main/java/org/wordpress/android/ui/posts/mediauploadcompletionprocessors/AudioBlockProcessor.kt

Utilized the local ID_ATTRIBUTE instead of the one from the FileBlockProcessor.

Co-authored-by: Matthew Kevins <mkevins@users.noreply.github.com>

guarani commented:
👋 Hi @jd-alexander, the Gutenberg editor release 1.44.0 is being cut now (apologies for the late headsup), and I came across this PR. If this doesn't depend on a specific Gutenberg editor release, and can just be picked up by the main apps release in 16.5, no action is needed. 
If this does depend on a Gutenberg Mobile PRs that are not currently tagged with the 1.44.0 milestone, but needs to make it into 16.5, please let me know and we'll merge it into the frozen release if needed.

jd-alexander commented:
I will merge this in when the [parent PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13612) is reviewed so that the diff of that PR isn't hard to read.

jd-alexander committed:
Merge branch 'develop' into gutenberg/audio-block-processor

jd-alexander committed:
synced submodules with develop so no changes contained in diff.

jd-alexander commented:
I am merging this because the [parent PR](https://github.com/wordpress-mobile/WordPress-Android/pull/13612) was mistakenly merged in and then reverted. Since this was already tested, it can be merged in, and when the parent PR is recreated it can branch from develop.
